src/Lykke.Service.CandlesHistory.Services/Candles/RedisCandlesCacheService.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Common;
+using JetBrains.Annotations;
+using Lykke.Job.CandlesProducer.Contract;
+using Lykke.Service.CandlesHistory.Core.Domain.Candles;
+using Lykke.Service.CandlesHistory.Core.Services.Candles;
+using MoreLinq;
+using StackExchange.Redis;
+
+namespace Lykke.Service.CandlesHistory.Services.Candles
+{
+    /// <summary>
+    /// Caches candles in the redis using lexographical indexes with candles data as the auxiliary information of the index
+    /// </summary>
+    [UsedImplicitly]
+    public class RedisCandlesCacheService : ICandlesCacheService
+    {
+        private const string TimestampFormat = "yyyyMMddHHmmss";
+
+        private readonly IDatabase _database;
+        private readonly MarketType _market;
+        private readonly int _amountOfCandlesToStore;
+
+        public RedisCandlesCacheService(IDatabase database, MarketType market, int amountOfCandlesToStore)
+        {
+            _database = database;
+            _market = market;
+            _amountOfCandlesToStore = amountOfCandlesToStore;
+        }
+
+        public IImmutableDictionary<string, IImmutableList<ICandle>> GetState()
+        {
+            throw new NotSupportedException();
+        }
+
+        public void SetState(IImmutableDictionary<string, IImmutableList<ICandle>> state)
+        {
+            throw new NotSupportedException();
+        }
+
+        public string DescribeState(IImmutableDictionary<string, IImmutableList<ICandle>> state)
+        {
+            throw new NotSupportedException();
+        }
+
+        public async Task InitializeAsync(
+            string assetPairId, 
+            CandlePriceType priceType,
+            CandleTimeInterval timeInterval,
+            IReadOnlyCollection<ICandle> candles)
+        {
+            foreach (var candle in candles)
+            {
+                if (candle.AssetPairId != assetPairId)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid AssetPriceId", nameof(candles));
+                }
+                if (candle.PriceType != priceType)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid PriceType", nameof(candles));
+                }
+                if (candle.TimeInterval != timeInterval)
+                {
+                    throw new ArgumentException($"Candle {candle.ToJson()} has invalid TimeInterval", nameof(candles));
+                }
+            }
+
+            // TODO: This is non concurrent-safe update
+
+            var key = GetKey(assetPairId, priceType, timeInterval);
+
+            // Cleans cache
+
+            await _database.SortedSetRemoveRangeByRankAsync(key, 0, -1);
```

Comment:
Why not just delete the key?

Reply:
Such an initialization is still used only for purposes of an old unit test. The test was originally designed for in-memory cache service and will be fully adopted for Redis in future. Anyway, there is no a real-working scenarion with manual initializatin of the cache in CandlesHistoryService (such an initialization is to be performed in Lykke.Job.CandlesHistoryWriter startup).